### PR TITLE
Spouse demo

### PR DESCRIPTION
**Pull request by brahmaneya**
1. Adapted Spouse tutorial to work on redux.
2. Changed end model to tensorflow / keras.

**Test Plan**
tox -e spouse and tox are passing.

### CODE REVIEW

File: spouse/preprocessors.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+from snorkel.labeling.preprocess import preprocessor
+from snorkel.types import DataPoint
+
+
+@preprocessor()
```
Comment (paroma) [inline spouse/preprocessors.py]: do we want typing in the tutorial? cc @henryre

File: spouse/preprocessors.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    cand.person2_left_tokens = cand.tokens[0:end][-1 - window : -1]
+    return cand
+
+
+@preprocessor()
+def get_right_tokens(cand: DataPoint) -> DataPoint:
+    """
+    Returns the length window tokens between to the right of the person mentions
+    """
+    # TODO: need to pass window as input params
```
Comment (paroma) [inline spouse/preprocessors.py]: update doctoring to start "length 3 window"

File: spouse/spouse_demo.py [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+        return NEGATIVE
+    elif len(set(family).intersection(set(x.person2_left_tokens))) > 0:
+        return NEGATIVE
+    else:
+        return ABSTAIN
+
+
+# %%
+# Check for `other` relationship words between person mentions
+other = {"boyfriend", "girlfriend", "boss", "employee", "secretary", "co-worker"}
```
Comment (paroma) [inline spouse/spouse_demo.py]: change from set to list to be consistent with above examples

File: spouse/spouse_demo.py [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+    p1, p2 = x.person_names
+    if (p1, p2) in known_spouses or (p2, p1) in known_spouses:
+        return POSITIVE
+    else:
+        return ABSTAIN
+
+
+# %%
+# Helper function to get last name for dbpedia entries.
+def last_name(s):
```
Comment (paroma) [inline spouse/spouse_demo.py]: maybe this should be hidden in utils.py? it might be confusing to see a function that looks like a preprocessor but isn't being used like one (since it operates over the resource not the data)

or maybe adding a note to point out the difference

File: spouse/spouse_demo.py [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+#     cardinality  Cardinality, or the number of classes in the task
+#     lr  The factor by which we update model weights after computing the gradient
+#     class_balance Proportion of [positive, negative] samples in the dataset
+#     n_epochs     A single pass through all the data in your training set
+
+# %%
+from snorkel.labeling.model.label_model import LabelModel
+
+label_model = LabelModel(cardinality=2, verbose=True)
+label_model.fit(train_L, dev_labels, l2=0.0, n_epochs=5000, log_freq=500, seed=123)
```
Comment (paroma) [inline spouse/spouse_demo.py]: l2=0.0 by default, is this to show the parameter exists?

File: spouse/spouse_demo.py [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+# ### Label Model Accuracy
+# This is the accuracy of the labels inferred by the Label Model.
+
+# %%
+from snorkel.analysis.metrics import metric_score
+from snorkel.analysis.utils import probs_to_preds
+
+Y_probs_dev = label_model.predict_proba(dev_L)
+Y_preds_dev = probs_to_preds(Y_probs_dev)
+metric_score(dev_labels, Y_preds_dev, probs=Y_probs_dev, metric="accuracy")
```
Comment (paroma) [inline spouse/spouse_demo.py]: maybe f1 score instead of accuracy because of the class imbalance?

File: spouse/spouse_demo.py [hunk 7]
```diff
@@ -0,0 +1,552 @@
...
+
+# %% [markdown]
+# ### Majority Vote Comparison
```
Comment (paroma) [inline spouse/spouse_demo.py]: we can probably remove this comparison

File: spouse/spouse_demo.py [hunk 8]
```diff
@@ -0,0 +1,552 @@
...
+# %% [markdown]
+# ## Part 4: Training our End Extraction Model
+#
+# In this final section of the tutorial, we'll use our noisy training labels alongside the development set labels to train our end machine learning model. We start by filtering out training examples which did not recieve a label from any LF, as these examples contain no signal. Then we concatenate them with dev set examples.
+#
+# %%
+import pandas as pd
+from snorkel.analysis.utils import preds_to_probs
+
+dev_label_probs = preds_to_probs(dev_labels, 2)
```
Comment (paroma) [inline spouse/spouse_demo.py]: add a note about changing to probs because of the end model

File: spouse/spouse_demo.py [hunk 9]
```diff
@@ -0,0 +1,552 @@
...
+    idx1 = np.array(list(map(list, df.person1_word_idx)))
+    idx2 = np.array(list(map(list, df.person2_word_idx)))
+    return tokens, idx1, idx2
+
+
+test_tokens, test_idx1, test_idx2 = get_feature_arrays(test_df)
+probs = model.predict((test_tokens, test_idx1, test_idx2))
+preds = probs_to_preds(probs)
+print(
+    f"Test Accuracy when trained with soft labels: {metric_score(test_labels, preds=preds, metric='accuracy')}"
```
Comment (paroma) [inline spouse/spouse_demo.py]: f1 instead of accuracy

File: spouse/spouse_demo.py [hunk 10]
```diff
@@ -0,0 +1,552 @@
...
+)
+print(
+    f"Test ROC-AUC when trained with soft labels: {metric_score(test_labels, probs=probs, metric='roc_auc')}"
+)
+
+# %% [markdown]
+# We compare this to an equivalent model that is only trained using the dev set labels. The accuracy is similar in other cases, because the dataset is very unbalanced. However, the ROC-AUC is lower when just training on the dev set, due to the much smaller number of examples.
+
+# %%
+dev_label_probs = preds_to_probs(dev_labels, 2)
```
Comment (paroma) [inline spouse/spouse_demo.py]: was `dev_label_probs ` reassigned? remove if repeated assignment

File: spouse/spouse_demo.py [hunk 11]
```diff
@@ -0,0 +1,552 @@
...
+    (dev_tokens, dev_idx1, dev_idx2),
+    dev_label_probs,
+    batch_size=batch_size,
+    epochs=num_epochs,
+)
+
+dev_probs = dev_model.predict((test_tokens, test_idx1, test_idx2))
+dev_preds = probs_to_preds(dev_probs)
+print(
+    f"Test Accuracy when trained with dev labels: {metric_score(test_labels, preds=dev_preds, metric='accuracy')}"
```
Comment (paroma) [inline spouse/spouse_demo.py]: f1

File: spouse/spouse_demo.py (hunk 3 above)
Comment (brahmaneya) [reply spouse/spouse_demo.py]: We need sets to compute intersection. So I changed the above (family) to be a set instead.

File: spouse/spouse_demo.py (hunk 5 above)
Comment (brahmaneya) [reply spouse/spouse_demo.py]: Removed l2.

File: spouse/spouse_demo.py (hunk 9 above)
Comment (brahmaneya) [reply spouse/spouse_demo.py]: F1 score is always zero, because the model always assigns more than 0.5 probability to negative (due to the dataset imbalance). So we use ROC-AUC to measure model quality (the baseline that always outputs negative gets ROC-AUC of 0.5).

File: spouse/spouse_demo.py (hunk 11 above)
Comment (brahmaneya) [reply spouse/spouse_demo.py]: Same as above

File: spouse/spouse_demo.py (hunk 9 above)
Comment (paroma) [reply spouse/spouse_demo.py]: TODO @paroma to tune LabelModel to fix this

File: .gitignore [hunk 12]
```diff
@@ -124,6 +124,7 @@ dmypy.json
 
 # Editors
 .vscode/
+*.swp
```
Comment (vincentschen) [inline .gitignore]: nit: can we move these to a separate PR?

File: spouse/.gitignore [hunk 13]
```diff
@@ -0,0 +1,3 @@
+# Stored data
+*.pkl
+trained_spouse_model
```
Comment (vincentschen) [inline spouse/.gitignore]: nit: is there a way to make this consistent with the other checkpoints of other notebooks? I believe there's a `.gitignore` for `checkpoints` in master: https://github.com/snorkel-team/snorkel-tutorials/blob/master/.gitignore#L137

File: .gitignore (hunk 12 above)
Comment (vincentschen) [inline .gitignore]: nit: can we move these to a separate PR?

File: spouse/spouse_demo.py [hunk 14]
```diff
@@ -0,0 +1,363 @@
+# %% [markdown]
+# # Detecting spouse mentions in sentences
+
+# %% [markdown]
+# We will walk through an example text classification task to explore how Snorkel can be used for Information Extraction.
+# ## Classification Task
+# <img src="imgs/sentence.jpg" width="700px;">
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: fyi—this breaks in my browser preview: 
image

File: spouse/spouse_demo.py [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+#
+# We want to classify each __candidate__ or pair of people mentioned in a sentence, as being married at some point or not.
+#
+# In the above example, our candidate represents the possible relation `(Barack Obama, Michelle Obama)`. As readers, we know this mention is true due to external knowledge and the keyword of `wedding` occuring later in the sentence.
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: I'm assuming this is referring to the image?

File: spouse/spouse_demo.py [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+
+if os.path.basename(os.getcwd()) == "snorkel-tutorials":
+    os.chdir("spouse")
+
+from utils import load_data
+
+((df_dev, Y_dev), df_train, (df_test, Y_test)) = load_data()
+
+# %% [markdown]
+# **Input Data:** `df_dev` is a Pandas DataFrame object, where each row represents a particular __candidate__. The DataFrames contain the fields `sentence`, which refers to the sentence the candidate is in, `tokens`, the tokenized form of the sentence, `person1_word_idx` and `person2_word_idx`, which represent `[start, end]` indices in the tokens at which the first and second person's name appear, respectively.
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: Can we briefly explain the key properties of a candidate?

File: spouse/spouse_demo.py [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+
+candidate = df_dev.loc[2]
+person_names = get_person_text(candidate).person_names
+
+print("Sentence: ", candidate["sentence"])
+print("Person 1: ", person_names[0])
+print("Person 2: ", person_names[1])
+
+# %% [markdown]
+# ### Preprocessing the Database
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: Was database previously mentioned? Maybe change the wording here?

File: spouse/spouse_demo.py [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+#
+# In a real application, there is a lot of data preparation, parsing, and database loading that needs to be completed before we generate candidates and dive into writing labeling functions. Here we've pre-generated candidates in a pandas DataFrame object per split (train,dev,test).
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: Let's not overload "operators", because that's what we've been calling LFs/TFs/SFs.
Maybe we can allude to these as intuitive patterns, or something to that effect.

File: spouse/spouse_demo.py [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+    start = cand.person1_word_idx[1] + 1
+    end = cand.person2_word_idx[0]
+    cand.text_between = " ".join(cand.tokens[start:end])
+    return cand
+
+
+# %% [markdown]
+# ### Candidate PreProcessors
+#
+# We provide a set of `preprocessor`s for this task in `preprocessors.py`. For the purpose of the tutorial, we also have three fields (between_tokens, person1_right_tokens, person2_right_tokens) preprocessed in the data, which can be used when creating labeling functions.
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: nit: can we format the "three fields" using backticks to show that they are code properties? e.g. `between_tokens`

File: spouse/spouse_demo.py [hunk 20]
```diff
@@ -0,0 +1,363 @@
...
+#
+# `get_person_text(cand)`
+#
+# `get_person_lastnames(cand)`
+#
+# `get_left_tokens(cand)`
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: maybe a clearer way to frame this section is to enumerate/itemize all the fields that are available now, whether they come from preprocessors or are available beforehand

File: spouse/spouse_demo.py [hunk 21]
```diff
@@ -0,0 +1,363 @@
+# %% [markdown]
+# # Detecting spouse mentions in sentences
+
+# %% [markdown]
+# We will walk through an example text classification task to explore how Snorkel can be used for Information Extraction.
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: can we highlight up front the deltas from this to the intro spam tutorial? (we are pulling from even more labeling sources, this is a different task, etc.)

File: spouse/spouse_demo.py [hunk 22]
```diff
@@ -0,0 +1,363 @@
...
+Y_preds_dev = probs_to_preds(Y_probs_dev)
+print(
+    f"Label model f1 score: {metric_score(Y_dev, Y_preds_dev, probs=Y_probs_dev, metric='f1')}"
+)
+print(
+    f"Label model roc-auc: {metric_score(Y_dev, Y_preds_dev, probs=Y_probs_dev, metric='roc_auc')}"
+)
+
+# %%
+import matplotlib.pyplot as plt
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: Can we add a description about why we are showing this plot?

File: spouse/spouse_demo.py [hunk 23]
```diff
@@ -0,0 +1,363 @@
...
+    X=df_train, y=Y_probs_train, L=train_L
+)
+
+df_combined = pd.concat([df_dev, df_train_filtered])
+Y_probs_combined = np.concatenate([Y_probs_dev, Y_probs_train_filtered], 0)
+
+# %% [markdown]
+# ### II. Training a _Long Short-term Memory_ (LSTM) Neural Network
+#
+# [LSTMs](https://en.wikipedia.org/wiki/Long_short-term_memory) can acheive state-of-the-art performance on many text classification tasks. We'll train a simple LSTM model below. tf_model contains functions for processing features and building the tensorflow graphs for training and evaluation.
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: is this necessary? who is our audience here?

Comment (vincentschen) [reply spouse/spouse_demo.py]: nit: format `tf_model` with backticks

File: spouse/spouse_demo.py [hunk 24]
```diff
@@ -0,0 +1,363 @@
...
+#
+# For this tutorial, we will be training a fairly effective deep learning model. More generally, however, Snorkel plugs in with many ML libraries, making it easy to use almost any state-of-the-art model as the end model!
+
+# %%
+from tf_model import get_model
+
+
+# Truncate sentences to limit memory usage when padding.
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: add a a comment about what this feature means?

File: spouse/spouse_demo.py [hunk 25]
```diff
@@ -0,0 +1,363 @@
...
+tokens, idx1, idx2 = get_feature_arrays(df_combined)
+
+batch_size = 64
+num_epochs = 20  # TODO: Change this to ~50. Warning: Training takes several minutes!
+model.fit(
+    (tokens, idx1, idx2), Y_probs_combined, batch_size=batch_size, epochs=num_epochs
+)
+
+# %% [markdown]
+# Measure the trained model's prediction F1 score and ROC_AUC.
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: nit: can we keep using the first person here?

File: spouse/spouse_demo.py [hunk 26]
```diff
@@ -0,0 +1,363 @@
...
+# %%
+test_tokens, test_idx1, test_idx2 = get_feature_arrays(df_test)
+probs = model.predict((test_tokens, test_idx1, test_idx2))
+preds = probs_to_preds(probs)
+print(
+    f"Test F1 when trained with soft labels: {metric_score(Y_test, preds=preds, metric='f1')}"
+)
+print(
+    f"Test ROC-AUC when trained with soft labels: {metric_score(Y_test, probs=probs, metric='roc_auc')}"
+)
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: Can we add a few conclusion sentences about what we've achieved, etc. so the takeaways/gains are crystal clear?

File: spouse/spouse_demo.py (hunk 14 above)
Comment (brahmaneya) [reply spouse/spouse_demo.py]: Fixed. The image was empty for some reason.

File: spouse/spouse_demo.py (hunk 22 above)
Comment (brahmaneya) [reply spouse/spouse_demo.py]: Removed it. I had already removed the explanatory text earlier but forgot to remove this.

File: spouse/spouse_demo.py (hunk 24 above)
Comment (brahmaneya) [reply spouse/spouse_demo.py]: Moved to tf_model, and added types and docstring.

File: spouse/spouse_demo.py [hunk 27]
```diff
@@ -0,0 +1,341 @@
...
+
+if os.path.basename(os.getcwd()) == "snorkel-tutorials":
+    os.chdir("spouse")
+
+from utils import load_data
+
+((df_dev, Y_dev), df_train, (df_test, Y_test)) = load_data()
+
+# %% [markdown]
+# **Input Data:** `df_dev`, `df_train`, and `df_test` are Pandas DataFrame objects, where each row represents a particular __candidate__. For our problem, a candidate consists of a sentence, and two people mentioned in the sentence. The DataFrames contain the fields `sentence`, which refers to the sentence of the candidate, `tokens`, the tokenized form of the sentence, and `person1_word_idx` and `person2_word_idx`, which represent `[start, end]` indices in the tokens at which the first and second person's name appear, respectively.
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: nit: can we format `Pandas DataFrame`?

File: spouse/spouse_demo.py [hunk 28]
```diff
@@ -0,0 +1,341 @@
...
+# %% [markdown]
+# ### Part 4: Training our End Extraction Model
+#
+# In this final section of the tutorial, we'll use our noisy training labels alongside the development set labels to train our end machine learning model. We start by filtering out training examples which did not recieve a label from any LF, as these examples contain no signal. Then we concatenate them with dev set examples.
+#
+# %%
+from snorkel.analysis.utils import preds_to_probs
+from snorkel.labeling.utils import filter_unlabeled_dataframe
+
+# Change dev labels 1D array to 2D probabilities array as required for training end model.
```
Comment (vincentschen) [inline spouse/spouse_demo.py]: nit: change "probabilities" -> "categorical labels" ?
